### PR TITLE
Optimize OTAP log transformation in the ClickHouse exporter

### PR DESCRIPTION
**Pull request by lquerel**
# Change Summary

Optimizes the ClickHouse exporter transformation path for canonical OTAP log batches.

The exporter now uses a specialized, shape-preserving transformer that produces the same ClickHouse schema and values as the existing generic transformation pipeline while avoiding its general-purpose overhead.

Unsupported OTAP layouts automatically fall back to the existing generic transformer. OTLP logs, traces, and other payloads continue to use the generic path.

This PR is stacked on #3709.

## Performance impact

The change was benchmarked against #3709 using:

- 8,192-log batches
- synchronous ClickHouse inserts
- `max_in_flight: 10`
- one df_engine core
- six ClickHouse cores
- twelve traffic-generator cores
- ClickHouse 25.6
- three 60-second repetitions per scenario

Median results for canonical DFE OTAP logs:

| Metric | #3709 baseline | This change | Impact |
| --- | --- | --- | --- |
| Written throughput at maximum load | 682,597 logs/s | 885,655 logs/s | +29.7% / 1.30x |
| DFE CPU at approximately 100,000 logs/s | 28.46% | 20.72% | -27.2% |

Written throughput at the fixed offered load remained approximately 100,000 logs/s, as expected.

These results indicate that canonical OTAP log workloads can gain approximately 30% maximum throughput while using approximately 27% less DFE CPU at 100,000 logs/s. Actual gains depend on the input layout, workload, and ClickHouse capacity. Inputs that do not match the canonical layout use the generic fallback path.

## What issue does this PR close?

- Related to #3512
- Implements the OTAP transformation optimization identified by the ClickHouse exporter benchmarks in #3512

## How are these changes tested?

Automated tests verify:

- exact output parity between the specialized and generic transformers
- logs with complete and omitted attribute sets
- multiple resources and scopes with mixed content
- service-name extraction, ID encoding, body conversion, and attribute maps
- automatic fallback when the canonical logs payload is unavailable
- dictionary-encoded attribute parent IDs
- null, missing, scalar, map, and slice attribute values
- independent specialized-path and fallback telemetry counters

## Are there any user-facing changes?

Yes. Canonical OTAP log batches are transformed more efficiently before insertion into ClickHouse.

There is no new configuration and no expected change to the generated ClickHouse schema or values. Unsupported input layouts automatically use the existing generic transformer.

### Changelog

- [X] Added a `.chloggen/*.yaml` entry
- [ ] This PR is a chore (indicated in title)
- [ ] This is a documentation-only PR.

### CODE REVIEW

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs [hunk 1]
```diff
@@ -330,7 +340,30 @@ impl Exporter<OtapPdata> for ClickhouseExporter {
...
+                                self.ch_metrics.record_log_fast_path();
+                                Ok(HashMap::from([(ArrowPayloadType::Logs, batch)]))
+                            }
+                            Ok(LogsFastTransform::NotApplicable(reason)) => {
+                                self.ch_metrics.record_log_transform_fallback();
+                                otap_df_telemetry::otel_debug!(
+                                    "clickhouse.exporter.transform.fallback",
+                                    message = "Using generic ClickHouse transform for OTAP logs.",
+                                    reason = reason,
+                                );
```
[the remark this answers is not in the document]
Comment (albertlockett) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs]: It's at debug level. I feel a certain amount of noise can be expected.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs [hunk 2]
```diff
@@ -330,7 +340,30 @@ impl Exporter<OtapPdata> for ClickhouseExporter {
...
+                                batch_transformer.apply_plan(arrow_records)
+                            }
+                            Err(error) => Err(error),
+                        }
+                    } else {
+                        batch_transformer.apply_plan(arrow_records)
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs:361]: Do we also need to increment `log_transform_fallback_batches ` here as well if the signal type was logs?

File: rust/otap-dataflow/crates/contrib-nodes/benches/exporters/clickhouse_exporter/logs_transformer.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            let mut writer = StreamWriter::try_new(&mut bytes, output.schema_ref())
+                .expect("create ArrowStream writer");
+            writer.write(&output).expect("encode ClickHouse batch");
+            writer.finish().expect("finish ArrowStream");
+            black_box(bytes)
+        });
+    });
+
+    group.finish();
+}
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/benches/exporters/clickhouse_exporter/logs_transformer.rs:53]: didn't see these bench results posted elsewhere , so I'll post the results from my machine for posterity:

```
clickhouse_logs_transform/generic/8192
                        time:   [513.13 µs 514.31 µs 515.46 µs]
                        thrpt:  [15.893 Melem/s 15.928 Melem/s 15.965 Melem/s]
                        
clickhouse_logs_transform/specialized/8192
                        time:   [268.83 µs 270.02 µs 271.37 µs]
                        thrpt:  [30.187 Melem/s 30.339 Melem/s 30.473 Melem/s]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
  
clickhouse_logs_transform/arrow_stream/8192
                        time:   [26.009 µs 26.050 µs 26.097 µs]
                        thrpt:  [313.91 Melem/s 314.47 Melem/s 314.97 Melem/s]
Found 4 outliers among 100 measurements (4.00%)
```

pretty nice result on the fast path

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+//! Specialized transformation of canonical OTAP log batches into ClickHouse columns.
+//!
+//! # Applicability
+//!
+//! The exporter attempts this fast path only for log signals whose original format is
+//! `SignalFormat::OtapRecords`. It applies when a `Logs` batch is present and its optional
+//! `resource`, `scope`, and `body` columns use the canonical struct layout. Parent IDs must be
+//! `UInt16` when the corresponding resource, scope, or log attribute payload is present. Missing
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:11]: So in cases  where the body, resource, etc. are not structs) or when the IDs are not u16, the batch would actually not be valid a OTAP batch. I suppose it doesn't hurt to use the fallback transform to preserve the current behaviour, but I wonder if we should just be rejecting these invalid batches and saving ourselves some work?

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs]: I will improve this.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs [hunk 5]
```diff
@@ -182,6 +185,170 @@ fn parent_ids_as_u16(batch: &RecordBatch) -> Result<Option<UInt16Array>, Clickho
...
+            }
+            value if value == AttributeValueType::Bool as u8 => {
+                if let Some(value) = self.booleans.and_then(|a| a.value_at(row)) {
+                    builder
+                        .values()
+                        .append_value(if value { "true" } else { "false" });
+                } else {
+                    builder.values().append_value("");
+                }
+            }
```
[the remark this answers is not in the document]
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs:262]: This is a false positive.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs]: In fact I will remove this log as we already have a metric.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+                }
+            },
+            None => None,
+        };
+
+        let resource_attrs = grouped_attributes(records, ArrowPayloadType::ResourceAttrs)?;
+        let scope_attrs = grouped_attributes(records, ArrowPayloadType::ScopeAttrs)?;
+        let log_attrs = records.get(ArrowPayloadType::LogAttrs);
+
+        let mut columns: Vec<(&'static str, ArrayRef)> = Vec::with_capacity(16);
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:173]: probably a pretty minor optimization, but we _could_ save ourselves a heap allocation per invocation if we kept this `columns` vec as a property of `LogTransformer` and cleared it each time `try_apply` is called

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs [hunk 7]
```diff
@@ -182,6 +185,170 @@ fn parent_ids_as_u16(batch: &RecordBatch) -> Result<Option<UInt16Array>, Clickho
...
+struct AttributeMapEncoder<'a> {
+    types: &'a UInt8Array,
+    keys: StringArrayAccessor<'a>,
+    strings: Option<StringArrayAccessor<'a>>,
+    integers: Option<Int64ArrayAccessor<'a>>,
+    doubles: Option<&'a Float64Array>,
+    booleans: Option<&'a BooleanArray>,
+    bytes: Option<&'a BinaryArray>,
+    serialized: Option<ByteArrayAccessor<'a>>,
+}
```
[the remark this answers is not in the document]
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs:197]: Fixed in ef07caf90

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs]: Fixed in ef07caf90

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+                        service_name = Some(value);
+                    }
+                }
+            }
+        }
+
+        map_builder.append(true)?;
+        if let Some(builder) = &mut service_builder {
+            builder.append_value(service_name.unwrap_or(""));
+        }
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:412]: Could be something we leave for a a future PR, but I think there might be a way to further improve the performance of this.

For each parent_id, we get the range of values from the input map, and copy them one by one into the new map builder.

We might be able to just use [`MutableArrayData`](https://arrow.apache.org/rust/arrow/array/struct.MutableArrayData.html) instead, like:

```rs
let keys_mutable = MutableArrayData::new(&[&keys.to_data()], true, keys.len());
let values_mutable = MutableArrayData::new(&[&values.to_data()], true, values.len());
let mut offsets_builder = Vec::with_capacity(parent_ids.len());
let mut total_attrs = 0;
     
for row in 0..parent_ids.len() {
        let compact_index = // ...

        if let Some(compact_index) = compact_index {
            let start = offsets[compact_index] as usize;
            let end = offsets[compact_index + 1] as usize;
            keys_mutable.extend(0, start, end);
            values_mutable.extend(0, start, end);
            offsets.push(total_attrs);
            total_attrs += end - start;
}

offsets.push(total_attrs);
let keys = make_array(keys_mutable.to_data());
let values = make_array(values_mutable.to_data());
let offset_buffer = Buffer::from_vec(std::mem::take(&mut self.offsets_builder));
```

Exercise for the agent to figure out how to convert these back into a `MapArray` :) The idea is then to basically create a struct array, etc. so we can call the `MapArray::try_new` constructor
https://docs.rs/arrow/latest/arrow/array/struct.MapArray.html#method.try_new
See also: https://docs.rs/arrow-array/59.2.0/src/arrow_array/builder/map_builder.rs.html#229-280

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs [hunk 9]
```diff
@@ -182,6 +185,170 @@ fn parent_ids_as_u16(batch: &RecordBatch) -> Result<Option<UInt16Array>, Clickho
...
+    let mut cursors = offsets[..offsets.len() - 1].to_vec();
+    let mut grouped_rows = vec![0_u32; attribute_parent_ids.len()];
+    for (row, id) in attribute_parent_ids.iter().enumerate() {
+        let Some(id) = id else {
+            continue;
+        };
+        let cursor = &mut cursors[id as usize];
+        grouped_rows[*cursor as usize] = row as u32;
+        *cursor += 1;
+    }
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs:328]: it might be a somewhat minor optimization, but as near as I can tell, the only place this is used is from within `LogsFastTransformer`?

We might be able to make this a method, and have these vecs be a properties on this struct that gets cleared between invocations. That way we cut down on a couple heap allocation? (or if we don't make this a method, we could pass mutable references to these vecs)

**Review by albertlockett — APPROVED** [review]
LGTM - a few questions/suggestions that shouldn't block this

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs (hunk 2 above)
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/mod.rs:361]: Fixed in `a12942920`.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs (hunk 4 above)
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:11]: For this PR, I retained the generic fallback to preserve the exporter's existing error and delivery behavior. Rejecting them earlier should be handled separately so we can define the validation boundary and verify failure accounting and `wait_for_result` behavior. This is therefore deferred to a correctness-focused follow-up rather than included in this performance PR.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs (hunk 6 above)
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:173]: I implemented and benchmarked this suggestion at a batch size of 8192 logs. The prototype caused an approximately 1.5% regression, so for I didn't retain this change.

It looks like, the saved allocation did not compensate for the additional `ArrayRef` ownership and cloning work required because `RecordBatch` consumes its output array vector. The current local vector remains faster for this workload.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs (hunk 8 above)
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/logs_fast.rs:412]: I implemented the complete `MutableArrayData` prototype, reconstructed the result with `MapArray::try_new`, and verified it with the existing equivalence and edge-case tests.

Unfortunately, I didn't observe any repeatable improvement.

For the current attribute cardinality, the setup and additional bookkeeping appear to outweigh the benefit of bulk range copying. It may be worth revisiting with a dedicated high-cardinality attribute benchmark. I will left a comment in the code to keep track of this idea.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs (hunk 9 above)
Comment (lquerel) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/clickhouse_exporter/transform/transform_attributes.rs:328]: I implemented a version of that but the allocation savings are smaller than the additional bookkeeping and code-generation cost in this workload, while owned local vectors produce the faster code path.
